src/ast/semantic_analyser.cpp

```diff
@@ -18,6 +18,24 @@ void SemanticAnalyser::visit(Integer &integer)
   integer.type = SizedType(Type::integer, 8);
 }
 
+void SemanticAnalyser::visit(PositionalParameter &param)
+{
+  param.type = SizedType(Type::integer, 8);
+  std::string pstr = bpftrace_.get_param(param.n);
+  if (is_final_pass()) {
+    if (!bpftrace_.is_numeric(pstr)) {
+      if (!call_ || call_->func != "str")
+        /*
```

Comment:
certainly there's some friction here with the visitor pattern, since it avoids exposing parent context.

the alternative would be for all Calls to iterate through and validate their arguments. I think we have enough runtime type information to do that:

```c++
void SemanticAnalyser::visit(Call &call)
{
  if (call.func != "str") {
    for (Expression *expr : *call.vargs) {
      if(PositionalParameter* posParam = dynamic_cast<PositionalParameter*>(expr)) {
        // …
      }
    }
  }
  // …
}
```

Reply:
Good idea, and I didn't like needing to create the call_ context. I looked at this and it starts to get messy -- it's not just calls that refer to positional parameters, it's also predicates and map keys, at least, so we'd be doing positional parameter checks in multiple places. With the current way I wrote it, we keep those checks in SemanticAnalyser::visit(PositionalParameter ..., which I think makes most sense.

Reply:
there's probably a way to generalize the shared responsibility (of calls, predicates and maps) into an interface (IManagesPositionalParameters). though I don't see a good way to make the visitor visit a superclass.

the call_ pointer really is so simple it's hard to beat. admittedly it reduces encapsulation and introduces extra statefulness to the visitor, but probably this is the entire point of having a Visitor class (to store useful state whilst we traverse).